src/java.base/share/classes/java/util/zip/InflaterInputStream.java

```diff
@@ -155,7 +155,12 @@ public int read(byte[] b, int off, int len) throws IOException {
                     reachEOF = true;
                     return -1;
                 }
-                if (inf.needsInput()) {
+                if (inf.needsInput() && !inf.hasPendingOutput()) {
+                    // Even if needsInput() is true, the native inflater may have some
+                    // buffered data which couldn't fit in to the output buffer during the
+                    // last call to inflate. Consume that bffered data first before calling
+                    // fill() to avoid an EOF error if no more input is available and the
+                    // next call to inflate will finish the inflation.
                     fill();
```

Comment:
We need to submit a bug against fill() too as it is not currently specified to throw when the input is at EOF.

Reply:
Not sure about this one? `fill()` is specified to throw an `IOException` and `EOFException` is an `IOException`. It probably depends on how you interpret the current "*if an I/O error has occurred*" description in the throws section.

Do you propose to be more specific there and write something like "*if an I/O error has occurred or the end of the input buffer has been reached*"?

Reply:
> Not sure about this one? `fill()` is specified to throw an `IOException` and `EOFException` is an `IOException`. It probably depends on how you interpret the current "_if an I/O error has occurred_" description in the throws section.
> 
> Do you propose to be more specific there and write something like "_if an I/O error has occurred or the end of the input buffer has been reached_"?

Like many JDK 1.1 era APIs, InflaterInputStream is very under-specified. In this case, the protected fill method is begging us to specify how it behaves when the input is at EOF. An I/O error is surprising here as it sets the protected len field to indicate EOF. We can't change it after 25 years so we will need to specify existing behavior.

Reply:
OK, I've opened [JDK-8292427: Improve specification of InflaterInputStream.fill()](https://bugs.openjdk.org/browse/JDK-8292427) to track this.